tracing_test.go

```diff
@@ -21,314 +21,120 @@
-			}
-
-			expected := TraceData{Annotations: tt.expected, BinaryAnnotations: binaryAnnotations, Source: source, Target: target, Method: "echo"}
-			assert.Equal(t, expected, state.call, "%v: Report args mismatch", tt.name)
-			curSpan := CurrentSpan(ctx)
-			assert.Equal(t, NewSpan(curSpan.TraceID(), curSpan.TraceID(), 0), state.span, "Span mismatch")
-		})
-	}
-}
-
-func TestTraceReportingDisabled(t *testing.T) {
-	var gotCalls int
-	testTraceReporter := TraceReporterFunc(func(_ TraceData) {
-		gotCalls++
-	})
-
-	traceReporterOpts := testutils.NewOpts().SetTraceReporter(testTraceReporter)
-	WithVerifiedServer(t, traceReporterOpts, func(ch *Channel, hostPort string) {
-		ch.Register(raw.Wrap(newTestHandler(t)), "echo")
-
-		ctx, cancel := NewContext(time.Second)
-		defer cancel()
-
-		CurrentSpan(ctx).EnableTracing(false)
-		_, _, _, err := raw.Call(ctx, ch, hostPort, ch.PeerInfo().ServiceName, "echo", nil, []byte("arg3"))
-		require.NoError(t, err, "raw.Call failed")
-
-		assert.Equal(t, 0, gotCalls, "TraceReporter should not report if disabled")
-	})
-}
-
-func TestTraceSamplingRate(t *testing.T) {
-	rand.Seed(10)
-
-	tests := []struct {
-		sampleRate  float64 // if this is < 0, then the value is not set.
-		count       int
-		expectedMin int
-		expectedMax int
-	}{
-		{1.0, 100, 100, 100},
-		{0.5, 100, 40, 60},
-		{0.1, 100, 5, 15},
-		{0, 100, 0, 0},
-		{-1, 100, 100, 100}, // default of 1.0 should be used.
-	}
-
-	for _, tt := range tests {
-		WithVerifiedServer(t, nil, func(ch *Channel, hostPort string) {
-			var reportedTraces int
-			testTraceReporter := TraceReporterFunc(func(_ TraceData) {
-				reportedTraces++
-			})
-
-			var tracedCalls int
-			testutils.RegisterFunc(ch, "t", func(ctx context.Context, args *raw.Args) (*raw.Res, error) {
-				if CurrentSpan(ctx).TracingEnabled() {
-					tracedCalls++
-				}
-
-				return &raw.Res{}, nil
-			})
-
-			opts := testutils.NewOpts().SetTraceReporter(testTraceReporter)
-			if tt.sampleRate >= 0 {
-				opts.SetTraceSampleRate(tt.sampleRate)
+		var response testtracing.TracingResponse
+		require.NoError(t, json.CallPeer(json.Wrap(ctx), peer, ch.PeerInfo().ServiceName,
+			"call", &testtracing.TracingRequest{}, &response))
+
+		// Spans are finished in inbound.doneSending() or outbound.doneReading(),
+		// which are called on different go-routines and may execute *after* the
+		// response has been received by the client. Give them a chance to finish.
+		for i := 0; i < 100; i++ {
+			if spanCount := len(testtracing.MockTracerSampledSpans(tracer)); spanCount == 2 {
+				break
 			}
-
-			client := testutils.NewClient(t, opts)
-			defer client.Close()
-
-			for i := 0; i < tt.count; i++ {
-				ctx, cancel := NewContext(time.Second)
-				defer cancel()
-
-				_, _, _, err := raw.Call(ctx, client, hostPort, ch.PeerInfo().ServiceName, "t", nil, nil)
-				require.NoError(t, err, "raw.Call failed")
+			time.Sleep(time.Millisecond) // max wait: 100ms
```

Comment:
yes, these are nearly identical tests, but they are using different encoding, and different encodings have different paths to invoke tracing functions from their call/handler files.